contracts/sentinel/PCVSentinel.sol

```diff
@@ -0,0 +1,180 @@
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                guardChecks[i] = true;
+                numGuardsWhoCanProtec++;
+            }
+        }
+
+        address[] memory guardsWhoCanProtec = new address[](numGuardsWhoCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(guardChecks[i]) {
+                numGuardsWhoCanProtec--;
+                guardsWhoCanProtec[numGuardsWhoCanProtec] = guards.at(i);
+            }
+        }
+
+        return guardsWhoCanProtec;
+    }
+
+    // ---------- Governor-or-Admin-Or-Guardian-Only State-Changing API ----------
+
+    /**
+     * @notice adds a guard contract to the PCV Sentinel
+     * @param guard the guard-contract to add
+     */
+    function knight(address guard) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.add(guard);
+
+        // Inform the kingdom of this glorious news
+        emit GuardAdded(guard);
+    }
+
+    /**
+     * @notice removes a guard
+     * @param traitor the guard-contract to remove
+     */
+    function slay(address traitor) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.remove(traitor);
+
+        // Inform the kingdom of this sudden and inevitable betrayal
+        emit GuardRemoved(traitor);
+    }
+
+    // ---------- Public State-Changing API ----------
+
+    /**
+     * @notice checks the guard and runs its protec actions if needed
+     * @param guard the guard to activate
+     */
+    function protec(address guard) 
+        external
+        override
+    {
+        require(guards.contains(guard), "Guard does not exist.");
+
+        bool needsProtec = IGuard(guard).check();
+
+        if (needsProtec) {
+            (address[] memory targets, bytes[] memory calldatas) = IGuard(guard).getProtecActions();
+
+            for(uint256 i=0; i<targets.length; i++) {
+                require(targets[i] != address(this), "Nyeh!");
+                (bool success, bytes memory data) = targets[i].call(calldatas[i]);
+                require(success, string(data)); // Fail this call if any of the sub-calls fail
+            }
+
+            emit Protected(guard);
+        } else {
+            revert("No need to protec.");
+        }
+    }
+
+    /**
+     * @notice activate many guards
+     * @dev we use (optional) try-catch here in case any guards fail, this could be the case
+     * if something changes before the transaction hits the chain\
+     * @param allowFailure whether or not the entire transaction should revert on failure
+     * @param whichGuards the guards to try to activate
+     */
+    function protecMany(bool allowFailure, address[] calldata whichGuards)
+        public 
+        override
+    {
+        for(uint256 i=0; i<whichGuards.length; i++) {
+            if (allowFailure) {
+                try this.protec(whichGuards[i]) {
```

Comment:
Why not turn `protec` into a `public` function and just call `try protec(whichGuards[i])`? 

Not generally a fan of calling smart contract methods from within the contract itself using `this.` syntax

Reply:
Ignore - needed to use `try`